src/Chapter2.hs

```diff
@@ -596,7 +623,9 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate xs = go xs []
+  where go [] accum = accum
+        go (y:ys) accum = go ys (accum ++ [y,y])
```

Comment:
Usually, adding elements to the end of the list is costly. Your implementation is correct :+1: But you can try to implement this function more efficiently by using `:` instead `++`. Maybe you can think of a few ways, how you can achieve such implementation? I would suggest pattern-matching on the list in `duplicate` directly, without the helper recursive function `go`.

Reply:
Good call, I'll try that out!